ddtrace/debugging/_origin/span.py

```diff
@@ -0,0 +1,248 @@
+        yield _frame
+        _frame = _frame.f_back
+
+
+@attr.s
+class EntrySpanProbe(LogFunctionProbe):
+    __span_class__ = "entry"
+
+    @classmethod
+    def build(cls, name: str, module: str, function: str) -> "EntrySpanProbe":
+        message = f"{cls.__span_class__} span info for {name}, in {module}, in function {function}"
+
+        return cls(
+            probe_id=str(uuid.uuid4()),
+            version=0,
+            tags={},
+            module=module,
+            func_qname=function,
+            evaluate_at=ProbeEvaluateTimingForMethod.ENTER,
+            template=message,
+            segments=[LiteralTemplateSegment(message)],
+            take_snapshot=True,
+            limits=DEFAULT_CAPTURE_LIMITS,
+            condition=None,
+            condition_error_rate=0.0,
+            rate=float("inf"),
+        )
+
+
+@attr.s
+class ExitSpanProbe(LogLineProbe):
+    __span_class__ = "exit"
+
+    @classmethod
+    def build(cls, name: str, filename: str, line: int) -> "ExitSpanProbe":
+        message = f"{cls.__span_class__} span info for {name}, in {filename}, at {line}"
+
+        return cls(
+            probe_id=str(uuid.uuid4()),
+            version=0,
+            tags={},
+            source_file=filename,
+            line=line,
+            template=message,
+            segments=[LiteralTemplateSegment(message)],
+            take_snapshot=True,
+            limits=DEFAULT_CAPTURE_LIMITS,
+            condition=None,
+            condition_error_rate=0.0,
+            rate=float("inf"),
+        )
+
+    @classmethod
+    def from_frame(cls, frame: FrameType) -> "ExitSpanProbe":
+        code = frame.f_code
+        return t.cast(
+            ExitSpanProbe,
+            cls.build(
+                name=code.co_qualname if sys.version_info >= (3, 11) else code.co_name,  # type: ignore[attr-defined]
+                filename=str(Path(code.co_filename).resolve()),
+                line=frame.f_lineno,
+            ),
+        )
+
+
+@dataclass
+class EntrySpanLocation:
+    name: str
+    start_line: int
+    end_line: int
+    file: str
+    module: str
+    probe: EntrySpanProbe
+
+
+class EntrySpanWrappingContext(WrappingContext):
+    def __init__(self, f):
+        super().__init__(f)
+
+        lines = linenos(f)
+        start_line = min(lines)
+        filename = str(Path(f.__code__.co_filename).resolve())
+        name = f.__qualname__
+        module = f.__module__
+        self.location = EntrySpanLocation(
+            name=name,
+            start_line=start_line,
+            end_line=max(lines),
+            file=filename,
+            module=module,
+            probe=t.cast(EntrySpanProbe, EntrySpanProbe.build(name=name, module=module, function=name)),
+        )
+
+    def __enter__(self):
+        super().__enter__()
+
+        root = ddtrace.tracer.current_root_span()
+        location = self.location
+        if root is None or root.get_tag("_dd.entry_location.file") is not None:
```

Comment:
java impl does something different, it add the tags to both the current active span and root span. not sure if we want to align that logic or not.